File: pkgs/python/opticontest/problems/salesman_euclid/brute_force.py
from opticontest.problems.salesman_euclid.common import Solution

class Algorithm:
    def __init__(self, points):
        self.points = points
        self.solution = None
        self.__rec([])

    def count(self):
        return len(self.points)

    def __rec(self, indices):
        if len(indices) == self.count():
            self.__add_solution(indices)
            return

        for index in [i for i in range(0, self.count()) if i not in indices]:
            self.__rec(indices + [index])

    def __add_solution(self, indices):
        new_solution = Solution(self.points, indices)
        if not self.solution or new_solution.length < self.solution.length:
            self.solution = new_solution

File: pkgs/ruby/lib/opticontest/problems/salesman_euclid/brute_force.rb
# frozen_string_literal: true

module Opticontest
  module Problems
    module SalesmanEuclid
      module BruteForce
        class Algorithm
          include Opticontest::Problems::SalesmanEuclid::Common

          attr_reader :points, :solution

          def initialize(points)
            raise TypeError unless points.is_a? Points

            @points = points
            @solution = nil
            rec []
          end

        private

          def rec(indices)
            return add_solution indices if indices.count == points.count

            ((0...points.count).to_a - indices).each do |index|
              rec indices + [index]
            end
          end

          def add_solution(indices)
            new_solution = Solution.new points, indices

            unless @solution.nil? || new_solution.length < @solution.length
              return
            end

            @solution = new_solution
          end
        end
      end
    end
  end
end
